### PR TITLE
arch/sim: Fix the up_cpu_start initialisation on OSX with SMP config enabled

### PR DESCRIPTION
**Pull request by sebastianene07**
## Summary

This PR fixes the issue described in #1708 . On Mac the `sem_init(..)` host call returns with error as unnamed semaphores are not supported. Because of this, the call to `up_cpu_start` fails and the simulation hangs.
The patch fixes this issue by using pthread_cond_t signalling instead of sempahores.

### CODE REVIEW

File: arch/sim/src/nuttx-names.in [hunk 1]
```diff
@@ -149,6 +149,9 @@ NXSYMBOLS(sched_yield)
 NXSYMBOLS(seekdir)
 NXSYMBOLS(select)
 NXSYMBOLS(sem_destroy)
+NXSYMBOLS(sem_open)
+NXSYMBOLS(sem_close)
+NXSYMBOLS(sem_unlink)
```
Comment (xiaoxiang781216) [inline arch/sim/src/nuttx-names.in:154]: let's order the list.

File: arch/sim/src/sim/up_simsmp.c [hunk 2]
```diff
@@ -53,7 +54,11 @@
 struct sim_cpuinfo_s
 {
   int cpu;    /* CPU number */
+#ifdef CONFIG_HOST_MACOS
+  sem_t *done;
+#else
   sem_t done; /* For synchronization */
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/up_simsmp.c:60]: should we switch to named semp for all platform? it's hard to maintain two codebase.

File: arch/sim/src/sim/up_simsmp.c [hunk 3]
```diff
@@ -324,11 +344,22 @@ int up_cpu_start(int cpu)
   /* Initialize the CPU info */
 
   cpuinfo.cpu = cpu;
-  ret = sem_init(&cpuinfo.done, 0, 0);
+
+#ifdef CONFIG_HOST_MACOS
+  done_sema = sem_open(g_sema_name_done, O_CREAT, 0644, 0);
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/up_simsmp.c:349]: The name need different for each cpu and for multiple nuttx instance

**Comment by xiaoxiang781216** [comment]
After more reviewing, I think we can remove the semphare from the code since the synchronization isn't really needed.

**Comment by sebastianene07** [comment]
> After more reviewing, I think we can remove the semphare from the code since the synchronization isn't really needed.

Ok I will remove it, in that case I will also drop the changes from `nuttx-names.in`

File: arch/sim/src/nuttx-names.in (hunk 1 above)
Comment (sebastianene07) [reply arch/sim/src/nuttx-names.in:154]: Ok I will order the list

File: arch/sim/src/sim/up_simsmp.c (hunk 2 above)
Comment (sebastianene07) [reply arch/sim/src/sim/up_simsmp.c:60]: I think it's a good idea

File: arch/sim/src/sim/up_simsmp.c (hunk 3 above)
Comment (sebastianene07) [reply arch/sim/src/sim/up_simsmp.c:349]: It's a good catch, thanks !

**Comment by yamt** [comment]
i suspect it's simpler to use pthread mutex/condvar.

**Comment by sebastianene07** [comment]
> i suspect it's simpler to use pthread mutex/condvar.

I updated the patch to use pthread_con_t signalling  mechanism

**Comment by xiaoxiang781216** [comment]
LGTM.

**Pull request merged**